pkg/backend/api/api_grpc_test.go

```diff
@@ -0,0 +1,112 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/mock/gomock"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+
+	accountclientmock "github.com/bucketeer-io/bucketeer/pkg/account/client/mock"
+	"github.com/bucketeer-io/bucketeer/pkg/cache"
+	cachev3mock "github.com/bucketeer-io/bucketeer/pkg/cache/v3/mock"
+	featureclientmock "github.com/bucketeer-io/bucketeer/pkg/feature/client/mock"
+	"github.com/bucketeer-io/bucketeer/pkg/log"
+	accountproto "github.com/bucketeer-io/bucketeer/proto/account"
+	environmentproto "github.com/bucketeer-io/bucketeer/proto/environment"
+)
+
+func TestCheckAuth(t *testing.T) {
+	t.Parallel()
+	mockController := gomock.NewController(t)
+	defer mockController.Finish()
+
+	patterns := []struct {
+		desc        string
+		setup       func(*PublicAPIService)
+		input       []accountproto.APIKey_Role
+		expected    *accountproto.EnvironmentAPIKey
+		expectedErr error
+	}{
+		{
+			desc: "error: invalid api key",
+			setup: func(s *PublicAPIService) {
+				s.environmentAPIKeyCache.(*cachev3mock.MockEnvironmentAPIKeyCache).EXPECT().Get(gomock.Any()).Return(
+					nil, cache.ErrNotFound)
+				s.accountClient.(*accountclientmock.MockClient).EXPECT().GetAPIKeyBySearchingAllEnvironments(gomock.Any(), gomock.Any()).Return(
+					nil, status.Errorf(codes.NotFound, "error: apy key not found"))
+			},
+			input: []accountproto.APIKey_Role{
+				accountproto.APIKey_PUBLIC_API_READ_ONLY,
+			},
+			expected:    nil,
+			expectedErr: ErrInvalidAPIKey,
+		},
+		{
+			desc: "success",
+			setup: func(gs *PublicAPIService) {
+				gs.environmentAPIKeyCache.(*cachev3mock.MockEnvironmentAPIKeyCache).EXPECT().Get(gomock.Any()).Return(
+					&accountproto.EnvironmentAPIKey{
+						Environment: &environmentproto.EnvironmentV2{Id: "ns0"},
+						ApiKey: &accountproto.APIKey{
+							Id:       "test-key",
+							Role:     accountproto.APIKey_PUBLIC_API_READ_ONLY,
+							Disabled: false,
+						},
+					}, nil)
+			},
+			input: []accountproto.APIKey_Role{
+				accountproto.APIKey_PUBLIC_API_READ_ONLY,
+			},
+			expected: &accountproto.EnvironmentAPIKey{
+				Environment: &environmentproto.EnvironmentV2{Id: "ns0"},
+				ApiKey: &accountproto.APIKey{
+					Id:       "test-key",
+					Role:     accountproto.APIKey_PUBLIC_API_READ_ONLY,
+					Disabled: false,
+				},
+			},
+			expectedErr: nil,
+		},
+	}
+
+	ctx := metadata.NewIncomingContext(context.TODO(), metadata.MD{
+		"authorization": []string{"test-key"},
+	})
+	for _, p := range patterns {
+		gs := newBackendServiceWithMock(t, mockController)
+		p.setup(gs)
+		actual, err := gs.checkAuth(ctx, p.input)
+		assert.Equal(t, p.expected, actual, "%s", p.desc)
+		assert.Equal(t, p.expectedErr, err, "%s", p.desc)
+	}
```

Comment:
I think `t.Run` makes the code cleaner and simple when displaying the test description.
Since we have many places using `t.Run`, it would be better to keep it for consistency.
WDYT?

```suggestion
	for _, p := range patterns {
		t.Run(p.desc, func(t *testing.T) {
			gs := newBackendServiceWithMock(t, mockController)
			p.setup(gs)
			actual, err := gs.checkAuth(ctx, p.input)
			assert.Equal(t, p.expected, actual)
			assert.Equal(t, p.expectedErr, err)
		}
	}
```